python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -751,6 +895,117 @@ def from_rule_evaluation_results(
+            if serialized_result[0].rule_snapshot != rule_snapshot:
+                continue
+            deserialized_result = self._deserialize_result(serialized_result, asset_graph)
+            if deserialized_result:
+                results.append(deserialized_result)
+        return results
+
+    def _get_asset_partitions_with_decision_type(
+        self, decision_type: AutoMaterializeDecisionType, asset_graph: AssetGraph
+    ) -> AbstractSet[AssetKeyPartitionKey]:
+        """Returns the set of asset partitions with a given decision type applied to them."""
+        asset_partitions = set()
+        for serialized_result in self.partition_subsets_by_condition:
+            if serialized_result[0].rule_snapshot.decision_type != decision_type:
+                continue
+            deserialized_result = self._deserialize_result(serialized_result, asset_graph)
+            if deserialized_result is None:
+                continue
+            asset_partitions.update(deserialized_result[1])
+        return asset_partitions
+
+    def get_requested_or_discarded_asset_partitions(
+        self, asset_graph: AssetGraph
+    ) -> AbstractSet[AssetKeyPartitionKey]:
+        """Returns the set of asset partitions which were either requested or discarded on this
+        evaluation.
+        """
+        to_materialize = self._get_asset_partitions_with_decision_type(
+            AutoMaterializeDecisionType.MATERIALIZE, asset_graph
+        )
+        if not to_materialize:
+            return set()
+        to_skip = self._get_asset_partitions_with_decision_type(
+            AutoMaterializeDecisionType.SKIP, asset_graph
+        )
+        return to_materialize - to_skip
+
+    def get_evaluated_asset_partitions(
+        self, asset_graph: AssetGraph
+    ) -> AbstractSet[AssetKeyPartitionKey]:
+        """Returns the set of asset partitions which were evaluated by any rule on this evaluation."""
+        # no asset partition can be evaluated by SKIP or DISCARD rules without having at least one
+        # materialize rule evaluation
+        return self._get_asset_partitions_with_decision_type(
+            AutoMaterializeDecisionType.MATERIALIZE, asset_graph
+        )
+
+    def equivalent_to_stored_evaluation(
+        self, stored_evaluation: Optional["AutoMaterializeAssetEvaluation"], asset_graph: AssetGraph
+    ) -> bool:
+        """This function returns if a stored record is equivalent to this one. To do so, we can't
+        just use regular namedtuple equality, as the serialized partition subsets will be
+        potentially have different string values.
+        """
+        if stored_evaluation is None:
+            return False
+        sorted_results = sorted(self.partition_subsets_by_condition)
+        sorted_stored_results = sorted(stored_evaluation.partition_subsets_by_condition)
+        return (
+            set(self.rule_snapshots or []) == set(stored_evaluation.rule_snapshots or [])
```

Comment:
If we add fields to `AutoMaterializeAssetEvaluation`, it seems like we it will be easy to forget to handle them here. Anything we can do to mitigate that?

Reply:
hm we could in theory just keep a set of "excluded fields" and "fields we need to check for equality in a weird way", then just do the normal equality check for the remainder of the fields, but for what it's worth, I think the `asset_key` / `partition_subsets_by_condition` / `rule_snapshots` essentially fully determine the equivalence of these two objects (the num_x are basically derived properties from those), so I think that might end up being a bit over-engineered for now

Reply:
Once we have deterministic serialization for partition subsets, do we need this? Or can we just compare the strings?

Reply:
ah missed this question -- the only remaining bit is that when this gets deserialized, partition_subsets_by_condition gets deserialized as a list of lists, rather than a list of tuples, but that could be fixed by overriding the __new__ on this, so yeah this could be totally removed I believe

Reply:
nice